core/src/banking_stage.rs

```diff
@@ -1853,57 +1866,100 @@ impl BankingStage {
         process_transactions_summary
     }
 
+    // Returns the number of packets that were filtered out for
+    // no longer being valid (could be too old, a duplicate of something
+    // already processed, etc.)
     fn filter_unprocessed_packets_at_end_of_slot(
-        bank: &Arc<Bank>,
-        deserialized_packet_batch: &HashMap<usize, DeserializedPacket>,
+        bank: &Option<Arc<Bank>>,
+        unprocessed_packets: &mut UnprocessedPacketBatches,
+        retryable_packets: &mut UnprocessedPacketBatches,
         my_pubkey: &Pubkey,
         next_leader: Option<Pubkey>,
         banking_stage_stats: &BankingStageStats,
-    ) -> Vec<usize> {
+    ) -> usize {
+        let (bigger_queue, smaller_queue, is_unprocessed_packets_bigger) =
+            if retryable_packets.len() > unprocessed_packets.len() {
+                (retryable_packets, unprocessed_packets, false)
+            } else {
+                (unprocessed_packets, retryable_packets, true)
+            };
+        let original_unprocessed_packets_len = bigger_queue.len() + smaller_queue.len();
+
         // Check if we are the next leader. If so, let's not filter the packets
         // as we'll filter it again while processing the packets.
         // Filtering helps if we were going to forward the packets to some other node
-        if let Some(leader) = next_leader {
-            if leader == *my_pubkey {
-                return deserialized_packet_batch
-                    .keys()
-                    .cloned()
-                    .collect::<Vec<usize>>();
+        let will_still_be_leader = next_leader
+            .map(|next_leader| next_leader == *my_pubkey)
+            .unwrap_or(false);
+        let should_filter_unprocessed_packets = !will_still_be_leader && bank.is_some();
+
+        if !should_filter_unprocessed_packets {
+            // If we shouldn't filter the packets, then simply merge the
+            // retryable queue and the unprocessed packets queue so we don't
+            // lose any packets for forwarding or for the next time we're
+            // processing transactions
+            for packet in smaller_queue.iter() {
+                bigger_queue.push(packet.clone());
```

Comment:
`MinMaxHeap`  implements `Extend`.  prolly faster and make the `clone` avoidable

Reply:
MinMaxHeap is one part, but the hashmap tracking the actual `DeserializedPacket` also needs to be extended. 

To do this, you could expose a `merge()` method between two `UnprocessedPacketBatch`'s, but  you'd also have to handle duplicate packets between the two. 

Rather not do that right now

Reply:
maybe add a comment for future, ambitious Carl?

Reply:
no longer needed because I deleted the need to merge UnprocessedPacketBatch's 😄